src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs

```diff
@@ -110,7 +110,27 @@ public async Task<ResourceDownloadResults> EnsureResource(IHostEnvironment env,
                     $"Could not create a valid URI from the base URI '{MlNetResourcesUrl}' and the relative URI '{relativeUrl}'");
             }
             return new ResourceDownloadResults(filePath,
-                await DownloadFromUrl(env, ch, absoluteUrl.AbsoluteUri, fileName, timeout, filePath), absoluteUrl.AbsoluteUri);
+                await DownloadFromUrlWithRetryAsync(env, ch, absoluteUrl.AbsoluteUri, fileName, timeout, filePath), absoluteUrl.AbsoluteUri);
+        }
+
+        private async Task<string> DownloadFromUrlWithRetryAsync(IHostEnvironment env, IChannel ch, string url, string fileName,
+            int timeout, string filePath, int retryTimes = 5)
+        {
+            var downloadResult = "";
+
+            for (int i = 0; i < retryTimes; ++i)
+            {
+                var thisDownloadResult = DownloadFromUrl(env, ch, url, fileName, timeout, filePath);
+
+                if (string.IsNullOrEmpty(await thisDownloadResult))
+                    return await thisDownloadResult;
+                else
```

Comment:
There shouldnt be a need to await twice on the same async task. I think it would be cleaner if there was a single await call and then use the cached results.

Reply:
yes, thanks, fixed

---
In reply to: [375625182](https://github.com/dotnet/machinelearning/pull/4776#discussion_r375625182) [](ancestors = 375625182)